docs: Token usage and machine limits (#3222)
canonical/ubuntu-pro-client: docs/explanations/pro_token_and_machine_usage.rst

.. _pro_token_and_machine_usage:

Machine counts and token usage for Ubuntu Pro
*********************************************

Ubuntu Pro tokens are used to authenticate your machine and enable access to the services you are entitled to. These tokens are unique and specific to your account, so you should treat them as a secret (as you would with a password).

In this page, we'll discuss how Ubuntu Pro tokens work and how machines are counted against your subscription.

> Ubuntu Pro token(s)
> --------------------

Reviewer: ```suggestion
About Ubuntu Pro tokens
-----------------------
```

- Obtaining Pro tokens:

>      After subscribing to Ubuntu Pro, you can access your subscriptions through the `Ubuntu Pro Dashboard <Pro_dashboard_>`_. Each subscription includes a Ubuntu Pro token that can be used to attach as many machines as the subscription allows.

Reviewer: ```suggestion
     After subscribing to Ubuntu Pro, you can access your subscriptions through the `Ubuntu Pro Dashboard <Pro_dashboard_>`_. Each subscription includes an Ubuntu Pro token that can be used to attach as many machines as your subscription allows.
```

>      After subscribing to Ubuntu Pro, you can access your subscriptions through the `Ubuntu Pro Dashboard <Pro_dashboard_>`_. Each subscription includes a Ubuntu Pro token that can be used to attach as many machines as the subscription allows.

Reviewer: Just looking at the rendered docs, the indent seems quite large. I think you only need two spaces worth of indentation for a bullet point in rst

- Attaching Pro tokens:

>     Command: Use ``sudo pro attach`` to connect a machine to your Ubuntu Pro subscription.

Reviewer: ```suggestion
    You can use the ``sudo pro attach`` command to connect a machine to your Ubuntu Pro subscription.
```

Reviewer: Having this after the "obtain a token" point makes it appear like this is a list of tasks that they need to do in order, but they don't need the token *a priori* to be able to attach. They only need the subscription - so maybe worth adding a line here to point out that the attach process allows them to choose the token they want to use?

Author: The attach basically either allows them to use their obtained token to attach or use `magic attach` which directs them to the pro dashboard where they use pro attach and the command takes them to the pro website. We can clarify this in the sentence.

- Detaching Pro tokens:

>     Command: Use ``sudo pro detach`` to remove a machine from the subscription. If you are attached on a VM you intend to delete, the token does not need to be detached before the VM is destroyed.

Reviewer: ```suggestion
    You can use the ``sudo pro detach`` command to remove a machine from the subscription. If you are attached on a VM you intend to delete, the token does not need to be detached before the VM is destroyed.
```

> For further information regarding pro commands, refer to our :ref:`Getting started-with Ubuntu Pro client tutorial<tutorial-commands>`.

Reviewer: ```suggestion
For further information regarding Pro commands, refer to our :ref:`Getting started with Ubuntu Pro client <tutorial-commands>` tutorial.
```

Monitoring active machines:
---------------------------

Active machines are counted based on the number of machines attached to your subscription at a given time. An active machine is defined as a machine that has been attached to a given subscription at any point within the past 24 hours.

> Each subscription type has a limit on the number of machines it can cover. For personal subscriptions, this limit is 5 physical machines with unlimited VMs on these machines. Enterprise subscriptions will have a higher limit and may differentiate between physical and virtual machines.

Reviewer: I wonder if these three options would look a bit clearer if presented as bullet points?

Author: I think the bullet points here break the continuity, currently all three points look better as paragraphs conveying different points about Active machines and related information. Dividing them into bullets would force some topic on each point which could be more confusing in my opinion.

Reviewer: That's fair - feel free to disregard this suggestion then

For use beyond 5 physical machines, consider an enterprise subscription.

> The Ubuntu Pro dashboard can help track the number of active machines, ensuring compliance with subscription limits. It provides a comprehensive view of all active and attached machines. It helps track the number of active machines, ensuring that you remain within the subscription limits.

Reviewer: The third sentence is a duplicate of the first (just reworded)

> Dealing with multiple VMs:

Reviewer: ```suggestion
Dealing with multiple VMs
```

Reviewer: We usually don't put colons at the end of a header

Reviewer: (in case I miss any)

---------------------------

> In an IaaS environment where node control is limited, you can contact `Canonical's sales team <pro_sales_form_>`_ to discuss purchasing licenses per single VM rather than per physical node.

Reviewer: What's the advantage to buying a license for a single VM? How do VMs and nodes relate to each other? I.e., if you have a license for a node, how many VMs might that include and therefore what's the advantage to buying an individual license (not sure if this info is in the scope of this document, but it's what I'm left wondering as a reader)

Author: I think some of these questions make sense but they might be more specific in nature, I would assume if there was a seperate document which explained VM licensing or even a sales/customer success document explaining these differences, it would be better?

Reviewer: I haven't seen anything like that, except what Lech and Rajan have mentioned in the FAQ replies :/ I'm not sure if that even exists

Author: Yeah the issue is, this information might be too specific and the FAQ's don't really talk about these in detail. @orndorffgrant any suggestions for this?

Reviewer: I think since we don't know of any good resources or references, we can note it down for future work to include it here. It's not crucial enough to hold up the PR over it.

Handling overage of active machines:
------------------------------------

> Canonical monitors the number of active machines. Occasional overages are generally acceptable, but persistent overages may lead to subscription termination.

Reviewer: I wonder if we need to expand this a little, to make it a bit less terse and discuss things in terms that I often see concerns/questions/anxieties expressed? There are many examples in the FAQ of people asking "what if I have X vms and I use the tokens..." so maybe it's worth using one of those examples as a hypothetical example to illustrate that we're not going to be knocking down their door if they accidentally have too many active machines? Something this vague is guaranteed to create more anxieties and send people to the FAQ to ask if their specific situation is going to get them in trouble/get their sub terminated.

Author: We can add a sub section about frequently asked questions about usage?
This can maybe hold these examples or atleast try to answer a more general concern about these active machines (based on the FAQs).

Reviewer: My aim would be to try and reflect some of the language that is used in the questions about VMs and machine counts, rather than including an actual FAQ (since we have one already). By mirroring the same language and answering the questions in a more friendly way (the current sentence could be read as slightly unfriendly).

Author: Will do

.. _pro_token_and_machine_usage:

> Machine counts and Token usage for Ubuntu Pro

Reviewer: We tend not to capitalise "token" - might want to do a global replace

*********************************************

> Pro Tokens:

Reviewer: ```suggestion
Pro Tokens
```

-----------

- Obtaining Tokens:

>     After subscribing to Ubuntu Pro, you can access your tokens through the Ubuntu Pro dashboard. Each token can be used to attach multiple machines depending on the subscription.

Reviewer: Let's link here to the Pro dashboard

- Attaching Tokens:

>     Command: Use ``sudo pro attach <token>`` to connect a machine to your Ubuntu Pro subscription.

Reviewer: ```suggestion
    Command: Use ``sudo pro attach`` to connect a machine to your Ubuntu Pro subscription.
```

- Detaching Tokens:

>     Command: Use ``sudo pro detach`` to remove a machine from the subscription.

Reviewer: Probably also worth mentioning here that users don't need to specifically detach a VM that they plan to destroy anyway, and that once they detach they can reuse the token

> Monitoring Active Machines:

Reviewer: ```suggestion
Monitoring active machines
```

Reviewer: Headings should follow sentence case (no need to capitalise the start of each word)

---------------------------

> 

Reviewer: Before anything else, we should define here what we mean by "active machines" (i.e., what we count as active)

- Machine Limits: 
    Each subscription type has a limit on the number of machines it can cover. For personal subscriptions, this limit is 5 physical machines with unlimited VMs on these machines. For enterprise subscriptions, the limit is based on the number of physical hosts and VMs.

- Active Machine Tracking: 

>     The Ubuntu Pro dashboard and tools like Landscape help track the number of active machines, ensuring compliance with subscription limits.

Reviewer: Lets link here to Landscape

- Unused Licenses and Active Users:

>     The Ubuntu Pro dashboard provides a comprehensive view of all active and attached machines. It shows the number of active machines and helps track unused licenses, ensuring that you remain within the subscription limits.

Reviewer: Probably should link to the part of the dashboard that shows this info

Reviewer: I think licenses is a Landscape-specific thing? So we might want to specify "Landscape licenses"

- Monitoring with Landscape:

>     Setup: Install the Landscape client and connect it to the Landscape server.

Reviewer: Might also want to provide a link here to some instructions about that. Possibly in the Landscape docs if they have a page that outlines that, and if not, the Pro handbook has content about getting started with Landscape

>     Audit Machines: Use Landscape’s remote script execution capability to run audit scripts, such as ``uastatus.sh``, to gather detailed information about your Ubuntu Pro subscription across all machines. This information can be exported into a CSV report for further analysis.

Reviewer: You will probably need to add CSV to the spelling exception list, the spell checker doesn't like it

Dealing with Multiple VMs:

> ---------------------------
> - VM License Management: 

Reviewer: ```suggestion
---------------------------

- VM License Management: 
```

>     In an IaaS environment where node control is limited, you can contact Canonical's sales team to discuss purchasing licenses per single VM rather than per physical node.

Reviewer: IaaS is also triggering the spell checker. Do we have a link for Sales contact details?

Author: Should be easy to find one and add it, no issues

> - Centralized Management: 

Reviewer: ```suggestion
- Centralised Management: 
```

>     For enterprise environments, using Landscape ensures all Ubuntu instances have the necessary Ubuntu Pro entitlements, helping identify any discrepancies or unauthorized usage.

Reviewer: ```suggestion
    For enterprise environments, using Landscape ensures all Ubuntu instances have the necessary Ubuntu Pro entitlements, helping identify any discrepancies or unauthorised usage.
```

Handling Overage of Active Machines:

> ------------------------------------
> Canonical monitors the number of active machines. Occasional overages are generally acceptable, but persistent overages may lead to subscription termination.

Reviewer: ```suggestion
------------------------------------

Canonical monitors the number of active machines. Occasional overages are generally acceptable, but persistent overages may lead to subscription termination.
```

Frequently Asked Questions:
===========================

- Can I use Ubuntu Pro Desktop for server use cases?
    No, using Ubuntu Pro Desktop for non-desktop use cases violates the terms of service. Use an Ubuntu Pro Server subscription for server use.

> - What if I need more than 5 machines for personal use?

Reviewer: I think this can be combined with the point about centralised management/Landscape so that it's not so much an FAQ but rather a point about small-scale vs large-scale use so that it fits better with the content on the page

    For extensive use beyond 5 machines, consider a commercial subscription. Ubuntu Pro offers scalable solutions for both personal and enterprise needs.

> - Is there a centralized tool for managing Ubuntu Pro subscriptions?

Reviewer: ```suggestion
- Is there a centralised tool for managing Ubuntu Pro subscriptions?
```

>     Yes, Landscape provides a centralized tool for managing Ubuntu Pro subscriptions, ensuring compliance and efficient management.

Reviewer: ```suggestion
    Yes, Landscape provides a centralised tool for managing Ubuntu Pro subscriptions, ensuring compliance and efficient management.
```

Reviewer: (this bit reads a bit sales-y, will want to rework this after we've decided on the general ordering)

Author: I honestly don't think we would need this given there is already a section that covers centralised management, this question just repeats the information in a different format

- How can I handle Ubuntu Pro in airgapped environments?

>     Use an airgapped contract server to distribute tokens within isolated environments, coordinated with Canonical’s Customer Success team.

Reviewer: We should also note that this requires a paid subscription and is not available on free subscriptions

> - How do I disable Ubuntu Pro notifications?

Reviewer: I'd remove this, it's not really relevant to the page

.. _pro_token_and_machine_usage:

Machine counts and token usage for Ubuntu Pro
*********************************************

Ubuntu Pro tokens are used to authenticate your machine and enable access to the services you are entitled to. tokens are unique and specific to your account, so you should treat them as a secret (as you would with a password).

> Since tokens are attached to your account and not your machine, you can re-use tokens, even multiple times per day. In this page, we'll discuss how tokens work and how machines are counted against your subscription.

Author: Yeah I agree, could help being more descriptive here

Pro tokens
----------

- Obtaining tokens:

>     After subscribing to Ubuntu Pro, you can access your tokens through the `Ubuntu Pro Dashboard <Pro_dashboard_>`_. Each token can be used to attach multiple machines depending on the subscription.

Reviewer 2: I think we should emphasize the 1:1 relationship between subscriptions and tokens.

Maybe something like this:
```suggestion
    After subscribing to Ubuntu Pro, you can access your subscriptions through the `Ubuntu Pro Dashboard <Pro_dashboard_>`_. Each subscription includes a token that can be used to attach as many machines as the subscription allows.
```

- Attaching tokens:
    Command: Use ``sudo pro attach`` to connect a machine to your Ubuntu Pro subscription.

> - Detaching tokens:

Reviewer 2: Detaching in general doesn't involve the token - maybe this section should be moved down to the Machine activity count section

Author: We could maybe just leave it here? It helps with the coherence of the topic of pro token related commands and the continuity of the document. This would look a little out of place in the machine activity count section since we don't talk about how detach affects machine counts. What do you think?

Reviewer: I agree, I think it makes more sense here

>     Command: Use ``sudo pro detach`` to remove a machine from the subscription. Once a machine is detached, the token can be reused to attach another machine. If you are attached on a VM you intend to delete, the token does not need to be detached before the VM is destroyed.

Reviewer 2: "Once a machine is detached, the token can be reused to attach another machine." implies that the detach is necessary to reuse the token, which isn't true. I think we can remove this sentence.

For further information regarding pro commands, refer to our :ref:`Getting started-with Ubuntu Pro client tutorial<tutorial-commands>`.

Monitoring active machines:
---------------------------

> Active machines are counted based on the number of machines attached to your subscription at a given time.

Reviewer 2: Let's add another sentence here: "An active machine is defined as a machine that has been attached to a given subscription at any point within the past 24 hours."

- Machine Limits: 

>     Each subscription type has a limit on the number of machines it can cover. For personal subscriptions, this limit is 5 physical machines with unlimited VMs on these machines. For enterprise subscriptions, the limit is based on the number of physical hosts and VMs. 

Reviewer 2: ```suggestion
    Each subscription type has a limit on the number of machines it can cover. For personal subscriptions, this limit is 5 physical machines with unlimited VMs on these machines. Enterprise subscriptions will have a higher limit and may differentiate between physical and virtual machines.
```

>     For extensive use beyond 5 machines, consider an enterprise subscription.

Reviewer 2: ```suggestion
    For use beyond 5 physical machines, consider an enterprise subscription.
```

Author: Unnecessary adjective there, my bad (blaming my Writing class in college) :sweat_smile:

- Active Machine Tracking: 

>     The Ubuntu Pro dashboard and tools like `Landscape <Landscape_>`_ help track the number of active machines, ensuring compliance with subscription limits.

Reviewer 2: The Pro dashboard is the authority here - I don't think we should mention landscape for this purpose.

- Unused Licenses and Active Users:

>     The Ubuntu Pro dashboard provides a comprehensive view of all active and attached machines. It shows the number of active machines and helps track unused Landscape licenses, ensuring that you remain within the subscription limits.

Reviewer 2: Why is this a separate bullet from the previous one?

In general I think we can stop using bullet points in this doc and put things into a few paragraphs

Author: We can do without bullet points here for sure

> - Monitoring with Landscape:
>     Setup: Install the Landscape client and connect it to the Landscape server. For instructions on how to install the Landscape client, refer to the `Landscape Installation Guide <Landscape_install_>`_.

Reviewer 2: I don't think this is really related to the purpose of this doc
```suggestion
```

>     Audit Machines: Use Landscape’s remote script execution capability to run audit scripts, such as ``uastatus.sh``, to gather detailed information about your Ubuntu Pro subscription across all machines. This information can be exported into a CSV report for further analysis.

Reviewer 2: I think we shuold limit this doc to the Pro subscription model's relation ship to tokens and machine activity, so all of the pieces that talk about using landscape to monitor things can be removed IMO
```suggestion
```

Author: Do we want to point towards Landscape for monitoring enterprise machines or would that be diverting from the main topic of the doc?

Reviewer: We might need to add Landscape stuff back in later (once we start moving towards adding this also to the Pro Handbook) but I think that can be done as future work if necessary.

Dealing with multiple VMs:
---------------------------

- VM License Management: 
    In an IaaS environment where node control is limited, you can contact `Canonical's sales team <pro_sales_form_>`_ to discuss purchasing licenses per single VM rather than per physical node.

> - Centralised Management: 
>     For enterprise environments, using Landscape ensures all Ubuntu instances have the necessary Ubuntu Pro entitlements, helping identify any discrepancies or unauthorised usage.

Reviewer 2: ```suggestion
```

Handling overage of active machines:
------------------------------------

Canonical monitors the number of active machines. Occasional overages are generally acceptable, but persistent overages may lead to subscription termination.

> Frequently Asked Questions:
> ===========================
> 
> - Can I use Ubuntu Pro Desktop for server use cases?
>     No, using Ubuntu Pro Desktop for non-desktop use cases violates the terms of service. Use an Ubuntu Pro Server subscription for server use.
> 
> - How can I handle Ubuntu Pro in airgapped environments?
>     Use an airgapped contract server to distribute tokens within isolated environments, coordinated with Canonical’s Customer Success team. Using Ubuntu Pro in airgapped environments requires a paid subscription and is not available for free subscriptions.

Reviewer 2: None of this seems particularly relevant to Pro tokens and machine activity.
```suggestion
```

Author: I was debating these being there too since we already point users to the FAQ page. This doc can do without these.
